internal/trie/node/branch_encode.go

```diff
@@ -0,0 +1,231 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package node
+
+import (
+	"bytes"
+	"fmt"
+	"hash"
+	"io"
+
+	"github.com/ChainSafe/gossamer/internal/trie/codec"
+	"github.com/ChainSafe/gossamer/internal/trie/pools"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// ScaleEncodeHash hashes the node (blake2b sum on encoded value)
+// and then SCALE encodes it. This is used to encode children
+// nodes of branches.
+func (b *Branch) ScaleEncodeHash() (encoding []byte, err error) {
+	buffer := pools.DigestBuffers.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer pools.DigestBuffers.Put(buffer)
+
+	err = b.hash(buffer)
+	if err != nil {
+		return nil, fmt.Errorf("cannot hash node: %w", err)
+	}
+
+	encoding, err = scale.Marshal(buffer.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("cannot scale encode hashed node: %w", err)
+	}
+
+	return encoding, nil
```

Comment:
I disagree, it's important to know which of the two possible errors was triggered. There would be a single error possible, then yes wrapping the error would have no point. What do you think?

On the other hand, I will change `node` to `branch` (and `leaf` for leaves) since we're handling a branch here!